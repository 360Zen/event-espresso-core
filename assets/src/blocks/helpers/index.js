/**
 * External dependencies
 */
import { shallow } from 'enzyme';
import { noop } from 'lodash';

/**
 * WordPress dependencies
 */
import {
	createBlock,
	getBlockType,
	registerBlockType,
} from '@wordpress/blocks';
import { BlockEdit } from '@wordpress/editor';

/**
 * copy of `/gutenberg/packages/block-library/src/test/helpers/index.js`
 * but using shallow rendering
 *
 * @param {string} name
 * @param {string} settings
 * @return {string} the rendered edit block
 */
export const blockEditRender = ( name, settings ) => {
	if ( ! getBlockType( name ) ) {
		registerBlockType( name, settings );
	}
	const block = createBlock( name );

	return shallow(
		<BlockEdit
			name={ name }
			isSelected={ false }
			attributes={ block.attributes }
			setAttributes={ noop }
			user={ {} }
		/>,
	);
};
