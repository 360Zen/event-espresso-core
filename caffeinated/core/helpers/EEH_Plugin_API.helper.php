<?php
/**
 * This file contains the EEH_Plugin_API helper class.
 * @package      Event Espresso
 * @subpackage helpers
 * @since           4.3.0
 */
if ( ! defined('EVENT_ESPRESSO_VERSION')) exit('No direct script access allowed');

/**
 * A helper class containing methods for easily registering COMPLETE components in the EE system.
 *
 * Used internally by EE for registering some caffeinated components, and provided as an easy
 * method for plugins to register entire components.  To be clear, there are certain systems in EE
 * where implementing them for a plugin requires multiple actions and filters being hooked into.
 * Within this class are wrappers for "registering" with a few parameters and then the registry
 * method
 * will take care of setting up all the required filters and hooks for that system to be successfully
 * hooked into.
 *
 * @package        Event Espresso
 * @subpackage  helpers
 * @since            4.3.0
 * @author          Darren Ethier
 */
class EEH_Plugin_API {

    /**
     * Holds registered EE_Admin_Pages
     * @var array
     */
    protected static $_ee_admin_page_registry = array();


    /**
     * Holds values for registered message types
     * @var array
     */
    protected static $_ee_message_type_registry = array();



    /**
     * holds values for registered messages shortcode libraries
     * @var array
     */
    protected static $_ee_messages_shortcode_registry = array();



    /**
     * The purpose of this method is to provide an easy way for addons to register their admin pages (
     * using the EE Admin Page loader system).
     *
     * @since 4.3.0
     *
     * @param  string $page_basename    This string represents the basename of the Admin Page
     *                                                        init.  The init file must use this basename in its name and
     *                                                        class (i.e. {page_basename}_Admin_Page_Init.core.php).
     * @param  string $page_path             This is the path where the registered admin pages reside (
     *                                                         used to setup autoloaders).
     * @param  array  $config                   An array of extra configuration options (optional) that will
     *                                                        be used in different circumstances (@todo this is currently
     *                                                        in flux, hence why we have an array so people can use
     *                                                        implemented options and at a later date we can add
     *                                                        additional ones without messing up existing usage)
     * @return void
     */
    public static function register_ee_admin_page( $page_basename, $page_path, $config = array() ) {

        if ( ! did_action( 'AHEE__EE_Admin__loaded' ) || did_action( 'AHEE__EE_System__initialize_last' )) {
            EE_Error::doing_it_wrong('EEH_Plugin_API::register_ee_admin_page', __('Should be only called on the "AHEE__EE_Admin__loaded" hook.','event_espresso'), '4.1' );
        }

        //add incoming stuff to our registry property
        self::$_ee_admin_page_registry[$page_basename] = array(
            'page_path' => $page_path,
            'config' => $config
            );

        //add filters but only if they haven't already been added otherwise we'll get duplicate filters added - not good.
        if ( ! has_filter( 'FHEE__EE_Admin_Page_Loader___get_installed_pages__installed_refs', array( 'EEH_Plugin_API', 'set_page_basename') ) ) {

            add_filter('FHEE__EE_Admin_Page_Loader___get_installed_pages__installed_refs', array( 'EEH_Plugin_API', 'set_page_basename' ), 10 );
            add_filter('FHEE__EEH_Autoloader__load_admin_core', array( 'EEH_Plugin_API', 'set_page_path' ), 10 );
        }

    }


    public static function set_page_basename( $installed_refs ) {
        foreach ( self::$_ee_admin_page_registry as $basename => $args ) {
            $installed_refs[] = $basename;
        }
        return $installed_refs;
    }



    public static function set_page_path( $paths ) {
        foreach ( self::$_ee_admin_page_registry as $basename => $args ) {
            $paths[] = $args['page_path'];
        }
        return $paths;
    }



    /**
     * Method for registering new message types in the EE_messages system.
     *
     * Note:  All message types must have the following files in order to work:
     *
     * - EE_Message_Template_Defaults extended class(es). (see /core/libraries/messages/defaults
     *     for examples).  Note, how there is a default class for each messenger/messagetype combo.
     *     This class is used for defining how the default templates get setup.
     * - Template files for default templates getting setup (or you can define it all in the above
     *      Default class).  See /core/libraries/messages/message_type/assets/defaults/ for examples.
     * - EE_Messages_Validator extended class(es).  See /core/libraries/messages/validators/email/
     *      for examples.  Note for any new message types, there will need to be a validator for each
     *      messenger combo this message type can activate with.
     * - And of course the main EE_{Message_Type_Name}_message_type class that defines the new
     *      message type and its properties.
     *
     * @since   4.3.0
     *
     * @param  array  $setup_args  {
     *        An array of arguments provided for registering the message type.
     *
     *        @type string $mtname                               Whatever is defined for the $name property of
     *                                                                          the message type you are registering (eg.
     *                                                                          declined_registration). Required.
     *        @type string $mtfilename                          The filename of the message type being
     *                                                                          registered.  This will be the main
     *                                                                          EE_{Messagetype_Name}_message_type class. (
     *                                                                          eg. EE_Declined_Registration_message_type.
     *                                                                          class.php). Required.
     *        @type array $autoloadpaths                       An array of paths to add to the messages
     *                                                                          autoloader for the new message type. Required.
     *        @type array $messengers_to_activate_with An array of messengers that this message
     *                                                                          type should activate with. Each value in the
     *                                                                          array should match the name property of a
     *                                                                          EE_messenger. Optional.
     *
     * }
     * @return void
     */
    public static function register_new_message_type( $setup_args = array()  ) {
        //make sure this was called in the right place!
        if ( ! did_action( 'EE_Brewing_Regular___messages_caf' ) || did_action( 'AHEE__EE_System__perform_activations_upgrades_and_migrations' )) {
            EE_Error::doing_it_wrong('EEH_Plugin_API::register_new_message_type', __('Should be only called on the "EE_Brewing_Regular__messages_caf" hook.','event_espresso'), '4.3.0' );
        }

        //required fields MUST be present, so let's make sure they are.
        if ( ! is_array( $setup_args ) || empty( $setup_args['mtname'] ) || empty( $setup_args['mtfilename'] ) || empty( $setup_args['autoloadpaths'] ) )
            throw new EE_Error( __( 'In order to register a message type with EEH_Plugin_API::register_new_message_type, you must include an array that contains the following keys: "mtname", "mtfilename", "autoloadpaths"', 'event_espresso' ) );

        $mtname = (string) $setup_args['mtname'];
        //setup $__ee_message_type_registry array from incoming values.
        self::$_ee_message_type_registry[$mtname] = array(
            'mtfilename' => (string) $setup_args['mtfilename'],
            'autoloadpaths' => (array) $setup_args['autoloadpaths'],
            'messengers_to_activate_with' => ! empty( $setup_args['messengers_to_activate_with'] ) ? (array) $setup_args['messengers_to_activate_with'] : array()
            );

        //add filters but only if they haven't already been added otherwise we'll get duplicate filters added - not good.
        if ( ! has_filter( 'FHEE__EE_Messages_Init__autoload_messages__dir_ref',  array( 'EEH_Plugin_API', 'register_msgs_autoload_paths') ) )
            add_filter( 'FHEE__EE_Messages_Init__autoload_messages__dir_ref', array( 'EEH_Plugin_API', 'register_msgs_autoload_paths'), 10 );

        if ( ! has_filter( 'FHEE__EE_messages__get_installed__messagetype_files', array( 'EEH_Plugin_API', 'register_messagetype_files') ) ) {
            add_filter('FHEE__EE_messages__get_installed__messagetype_files', array( 'EEH_Plugin_API', 'register_messagetype_files'), 10, 2 );
            add_filter( 'FHEE__EE_messenger__get_default_message_types__default_types', array( 'EEH_Plugin_API', 'register_messengers_to_activate_mt_with'), 10, 2 );
        }
    }



    /**
     * callback for FHEE__EE_messages__get_installed__messagetype_files filter.
     *
     * @since   4.3.0
     *
     * @param  array  $messagetype_files The current array of message type file names
     * @param  array  $type                      This is a string that indicates what type is requested.  It
     *                                                        could be either 'messengers', 'message_types', or 'all'.
     * @return  array                                 Array of message type file names
     */
    public static function register_messagetype_files( $messagetype_files, $type ) {
        if ( empty( self::$_ee_message_type_registry ) )
            return $messagetype_files;

        foreach ( self::$_ee_message_type_registry as $mt_reg ) {
            if ( empty( $mt_reg['mtfilename' ] ) )
                continue;
            $messagetype_files[] = $mt_reg['mtfilename'];
        }

        return $messagetype_files;
    }





    /**
     * callback for FHEE__EE_Messages_Init__autoload_messages__dir_ref filter.
     *
     * @since    4.3.0
     *
     * @param array $paths array of paths to be checked by EE_messages autoloader.
     * @return array
     */
    public static function register_msgs_autoload_paths( $paths  ) {

        //first message type registry autoloaders?
        if ( ! empty( self::$_ee_message_type_registry ) ) {
            foreach ( self::$_ee_message_type_registry as $mt_reg ) {
                if ( empty( $mt_reg['autoloadpaths'] ) )
                    continue;
                $paths = array_merge( $paths, $mt_reg['autoloadpaths'] );
            }
        }

        if ( !empty( self::$_ee_messages_shortcode_registry ) ) {
             foreach ( self::$_ee_messages_shortcode_registry as $st_reg ) {
                if ( empty( $st_reg['autoloadpaths'] ) )
                    continue;
                $paths = array_merge( $paths, $st_reg['autoloadpaths'] );
            }
        }

        return $paths;
    }





    /**
     * callback for FHEE__EE_messenger__get_default_message_types__default_types filter.
     *
     * @since   4.3.0
     *
     *
     * @param  array                $default_types array of message types activated with messenger (
     *                                                               corresponds to the $name property of message type)
     * @param  EE_messenger $messenger      The EE_messenger the filter is called from.
     * @return  array
     */
    public static function register_messengers_to_activate_mt_with( $default_types, EE_messenger $messenger ) {
        if ( empty( self::$_ee_message_type_registry ) )
            return $default_types;

        foreach ( self::$_ee_message_type_registry as $mtname => $mt_reg ) {
            if ( empty( $mt_reg['messengers_to_activate_with'] ) || empty( $mt_reg['mtfilenames'] ) )
                continue;

            //loop through each of the messengers and if it matches the loaded class then we add this message type to the
            foreach ( $mt_reg['messengers_to_activate_with'] as $msgr ) {
                if ( $messenger->name == $msgr ) {
                    $default_types[] = $mtname;
                }
            }
        }

        return $default_types;
    }






    /**
     * Helper method for registring a new shortcodes library class for the messages system.
     *
     * Note this is not used for adding shortcodes to existing libraries.  It's for registering anything
     * related to registering a new EE_{shortcode_library_name}_Shortcodes.lib.php class.
     *
     * @since    4.3.0
     *
     * @param  array  $setup_args {
     *       An array of arguments provided for registering the new messages shortcode library.
     *
     *       @type string $name                                         What is the name of this shortcode library
     *                                                                              (e.g. 'question_list');
     *       @type array  $autoloadpaths                          An array of paths to add to the messages
     *                                                                              autoloader for the new shortcode library
     *                                                                              class file.
     *       @type string $msgr_validator_callback            Callback for a method that will register the
     *                                                                              library with the messenger
     *                                                                              _validator_config. Optional.
     *       @type string $msgr_template_fields_callback  Callback for changing adding the
     *                                                                              _template_fields property for messenger.
     *                                                                              For example, the shortcode library may add
     *                                                                              a new field to the message templates.
     *                                                                              Optional.
     *       @type string $valid_shortcodes_callback         Callback for message types
     *                                                                              _valid_shortcodes array setup. Optional.
     *       @type array  $list_type_shortcodes                 If there are any specific shortcodes with this
     *                                                                             message shortcode library that should be
     *                                                                             considered "list type" then include them in an
     *                                                                             array.  List Type shortcodes are shortcodes that
     *                                                                             have a corresponding field that indicates how
     *                                                                             they are parsed. Optional.
     * }
     * @return void
     */
    public static function register_messages_shortcode_library( $name, $setup_args = array() ) {
        //make sure this was called in the right place!
        if ( ! did_action( 'EE_Brewing_Regular___messages_caf' ) || did_action( 'AHEE__EE_System__perform_activations_upgrades_and_migrations' )) {
            EE_Error::doing_it_wrong('EEH_Plugin_API::register_message_shortcode_library', __('Should be only called on the "EE_Brewing_Regular__messages_caf" hook.','event_espresso'), '4.3.0' );
        }

        //required fields MUST be present, so let's make sure they are.
        if ( ! is_array( $setup_args ) || empty( $setup_args['autoloadpaths'] ) ) {
            throw new EE_Error( __( 'In order to register a messages shortcode library with EEH_Plugin_API::register_messages_shortcode_library, you must include an array that contains the following key: "autoload_paths"', 'event_espresso' ) );
        }

        $name = (string) $name;
        self::$_ee_messages_shortcode_registry[$name] = array(
            'autoloadpaths' => (array) $setup_args['autoloadpaths'],
            'list_type_shortcodes' => !empty( $setup_args['list_type_shortcodes'] ) ? (array) $setup_args['list_type_shortcodes'] : array()
            );

         //add filters but only if they haven't already been added otherwise we'll get duplicate filters added - not good.
        if ( ! has_filter( 'FHEE__EE_Messages_Init__autoload_messages__dir_ref',  array( 'EEH_Plugin_API', 'register_msgs_autoload_paths') ) )
            add_filter( 'FHEE__EE_Messages_Init__autoload_messages__dir_ref', array( 'EEH_Plugin_API', 'register_msgs_autoload_paths'), 10 );

        //add below filters if the required callback is provided.
        if ( !empty( $setup_args['msgr_validator_callback'] ) )
            add_filter( 'FHEE__EE_messenger__get_validator_config', $setup_args['msgr_validator_callback'], 10, 2 );

        if ( !empty( $setup_args['msgr_template_fields_callback'] ) )
            add_filter( 'FHEE__EE_messenger__get_template_fields', $setup_args['msgr_template_fields_callback'], 10, 2 );

        if ( !empty( $setup_args['valid_shortcodes_callback'] ) )
            add_filter( 'FHEE__EE_Messages_Base__get_valid_shortcodes', $setup_args['valid_shortcodes_callback'], 10, 2 );

        if ( !empty( $setup_args['list_type_shortcodes'] ) )
            add_filter( 'FHEE__EEH_Parse_Shortcodes___parse_message_template__list_type_shortcodes', array( 'EEH_Plugin_API', 'register_list_type_shortcodes'), 10 );
    }



    /**
     * This is the callback for the FHEE__EEH_Parse_Shortcodes___parse_message_template__list_type_shortcodes
     * filter which is used to add additional list type shortcodes.
     *
     * @since 4.3.0
     *
     * @param  array $original_shortcodes
     * @return  array                                   Modifications to original shortcodes.
     */
    public static function register_list_type_shortcodes( $original_shortcodes ) {
        if ( empty( self::$_ee_messages_shortcode_registry ) )
            return $original_shortcodes;

        foreach ( self::$_ee_messages_shortcode_registry as $name => $sc_reg ) {
            if ( !empty( $sc_reg['list_type_shortcodes'] ) )
                $original_shortcodes = array_merge( $original_shortcodes, $sc_reg['list_type_shortcodes'] );
        }

        return $original_shortcodes;
    }

}
