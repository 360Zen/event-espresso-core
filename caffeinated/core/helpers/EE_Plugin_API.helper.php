<?php
/**
 * This file contains the EE_Plugin_API helper class.
 * @package 	  Event Espresso
 * @subpackage helpers
 * @since 		  4.4.0
 */
if ( ! defined('EVENT_ESPRESSO_VERSION')) exit('No direct script access allowed');

/**
 * A helper class containing methods for easily registering complete components in the EE system.
 *
 * Used internally by EE for registering some caffeinated components, and provide as an easy method
 * for plugins to register entire components.
 *
 * @package 		Event Espresso
 * @subpackage 	helpers
 * @since 			4.4.0
 * @author  		Darren Ethier
 */
class EE_Plugin_API {

}
