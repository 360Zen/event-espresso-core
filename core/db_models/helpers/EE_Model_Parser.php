<?php

if (!defined('EVENT_ESPRESSO_VERSION'))
	exit('No direct script access allowed');

/**
 * Event Espresso
 *
 * Event Registration and Management Plugin for WordPress
 *
 * @ package			Event Espresso
 * @ author			Seth Shoultes
 * @ copyright		(c) 2008-2011 Event Espresso  All Rights Reserved.
 * @ license			http://eventespresso.com/support/terms-conditions/   * see Plugin Licensing *
 * @ link					http://www.eventespresso.com
 * @ version		 	4.3
 *
 * ------------------------------------------------------------------------
 *
 * EE_Model_Parser
 *
 * @package			Event Espresso
 * @subpackage		
 * @author				Mike Nelson
 *
 * ------------------------------------------------------------------------
 */
class EE_Model_Parser {
	const table_alias_model_relation_chain_seperator = '__';
	const table_alias_model_relation_chain_prefix_end = '___';
	/**
	 * Adds a period onto the front and end of the string. This often helps in searching.
	 * For example, if we want to find the model name "Event", it can be tricky when the following are possible
	 * "","Event.EVT_ID","Event","Event_Venue.Venue.VNU_ID",etc. It's easier to look for ".Event." in 
	 * "..",".Event.EVT_ID.", ".Event.", and ".Event_Venue.Venue.VNU_ID", especially when the last example should NOT
	 * be found because the "Event" model isn't mentioned- it's just a string that has a model name that conicidentally
	 * has it as a substring
	 * @param string $string_to_pad
	 * @return string
	 */
	public static function pad_with_periods($string_to_pad){
		return ".".$string_to_pad.".";
	}
	/**
	 * Bsaically undoes _pad_with_periods
	 * @param string $string_to_trim
	 * @return string
	 */
	public static function trim_periods($string_to_trim){
		return trim($string_to_trim,'.');
	}
	
	/**
	 * Gets the calculatd table's alias 
	 * @param string $model_relation_chain or query param
	 * @return string which can be added onto table aliases to make them unique
	 */
	public static function extract_table_alias_model_relation_chain_prefix($model_relation_chain,$this_model_name){
		//eg $model_relation_chain = 'Venue.Event_Venue.Event.Registration", and $this_model_name = 'Event'
		$model_relation_chain = self::pad_with_periods($model_relation_chain);
		$this_model_name = self::pad_with_periods($this_model_name);
		//eg '.Venue.Event_Venue.Event.Registration." and '.Event.'
		//remove this model name and everything afterwards
		$pos_of_model_name = strpos($model_relation_chain,$this_model_name);
		$model_relation_chain = substr($model_relation_chain,0,$pos_of_model_name);
		//eg '.Venue.Event_Venue.'
		//trim periods
		$model_relation_chain = self::trim_periods($model_relation_chain);
		//eg 'Venue.Event_Venue'
		//replace periods with double-underscores
		$model_relation_chain = str_replace(".",self::table_alias_model_relation_chain_seperator,$model_relation_chain);
		//eg 'Venue__Event_Venue'
		if($model_relation_chain !=''){
			$model_relation_chain = $model_relation_chain.self::table_alias_model_relation_chain_prefix_end;
		}
		//eg 'Venue_Event_Venue___'
		return $model_relation_chain;
	}
	/**
	 * Gets the table's alias (without prefix or anything)
	 * @param string $table_alias_with_model_relation_chain_prefix which CAN have a table alias model relation chain prefix (or not)
	 * @return string
	 */
	public static function remove_table_alias_model_relation_chain_prefix($table_alias_with_model_relation_chain_prefix){
		//does this actually have a table alias model relation chain prefix?
		$pos = strpos($table_alias_with_model_relation_chain_prefix,self::table_alias_model_relation_chain_prefix_end);
		if(  $pos !== FALSE){
			//yes
			//find that triple underscore and remove it and everything before it
			$table_alias = substr($table_alias_with_model_relation_chain_prefix, $pos + strlen(self::table_alias_model_relation_chain_prefix_end));
		}else{
			$table_alias = $table_alias_with_model_relation_chain_prefix;
		}
		return $table_alias;
	}
	/**
	 * Gets the table alias model relation chain prefix from the table alias already containing it
	 * @param string $table_alias_with_model_relation_chain_prefix
	 * @return string
	 */
	public static function get_prefix_from_table_alias_with_model_relation_chain_prefix($table_alias_with_model_relation_chain_prefix){
		//does this actually have a table alias model relation chain prefix?
		$pos = strpos($table_alias_with_model_relation_chain_prefix,self::table_alias_model_relation_chain_prefix_end);
		if(  $pos !== FALSE){
			//yes
			//find that triple underscore and remove it and everything before it
			$prefix = substr($table_alias_with_model_relation_chain_prefix, 0, $pos + strlen(self::table_alias_model_relation_chain_prefix_end));
		}else{
			$prefix = '';
		}
		return $prefix;
	}
	/**
	 * Gets the table alias model relation chain prefix from teh query param
	 * @param type $query_param
	 * @param type $table_alias_sans_prefix
	 * @return string
	 */
	public static function extract_table_alias_model_relation_chain_from_query_param($query_param,$table_alias_sans_prefix){
		$pos = strpos($query_param,$table_alias_sans_prefix);
		if($pos == 0){
			return '';
		}
		$prefix = substr($query_param,0,$pos-1);
		//turn last . indicates the end of the prefix
		if($prefix){
			return str_replace(".",self::table_alias_model_relation_chain_seperator,$prefix).self::table_alias_model_relation_chain_prefix_end;
		}
		
	}
}

// End of file EE_Model_Parser.php