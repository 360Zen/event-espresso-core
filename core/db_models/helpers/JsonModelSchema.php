<?php
namespace EventEspresso\core\db_models\helpers;

use EEM_Base;
use EE_Model_Field_Base;
use EE_Model_Relation_Base;

defined('EVENT_ESPRESSO_VERSION') || exit;

/**
 * This is used to generate an array that can be used to generate a schema for a given model.
 * The format for the generated array follows the structure given in the json-schema standard
 * @see http://json-schema.org
 *
 * @package    EventEspresso
 * @subpackage core\db_models\helpers
 * @author     Darren Ethier
 * @since      4.9.24.rc.018
 */
class JsonModelSchema
{
    /**
     * Return the schema for a given model from a given model.
     * @param \EEM_Base $model
     * @return array
     */
    public function getModelSchema(EEM_Base $model)
    {
        $schema = $this->getInitialSchemaStructure($model);
        $fields_on_model = $model->field_settings();
        $relations_on_model = $model->relation_settings();
        $schema = array_merge($schema, $this->getModelSchemaForFields($fields_on_model));
        return array_merge($schema, $this->getModelSchemaForRelations($relations_on_model));
    }


    /**
     * Get the schema for a given set of model fields.
     * @param \EE_Model_Field_Base[]     $model_fields
     * @return array
     */
    public function getModelSchemaForFields(array $model_fields)
    {
        $schema = array();
        foreach ($model_fields as $field => $model_field) {
            if (! $model_field instanceof EE_Model_Field_Base) {
                continue;
            }
            $schema['properties'][$field] = $model_field->get_schema();
        }
        return $schema;
    }


    /**
     * Get the schema for a given set of model relations
     * @param EE_Model_Relation_Base[] $relations_on_model
     */
    public function getModelSchemaForRelations(array $relations_on_model)
    {
        $schema = array(
            'relations' => array(
                'type' => 'object',
                'description' => __('Relations for this model', 'event_espresso'),
                'properties' => array(),
                'readonly' => true
            )
        );
        foreach ($relations_on_model as $model_name => $relation) {
            if (! $relation instanceof EE_Model_Relation_Base) {
                continue;
            }
            $schema['relations']['properties'][$model_name] = $relation->get_schema();
        }
        return $schema;
    }


    /**
     * Outputs the schema header for a model.
     * @param \EEM_Base $model
     * @return array
     */
    public function getInitialSchemaStructure(EEM_Base $model)
    {
        return array(
            '$schema' => 'http://json-schema.org/draft-04/schema#',
            'title' => $model->get_this_model_name(),
            'type' => 'object',
            'properties' => array()
        );
    }


    /**
     * Helper function that simply allows one to get whatever has been generated in an actual json object as opposed to
     * an array.
     * @param array $array_to_convert
     * @return bool|false|mixed|string
     */
    public function toJson($array_to_convert)
    {
        return wp_json_encode($array_to_convert);
    }
}
