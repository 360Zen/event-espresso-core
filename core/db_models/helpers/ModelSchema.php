<?php
namespace EventEspresso\core\db_models\helpers;

use EEM_Base;
use EE_Model_Field_Base;

defined('EVENT_ESPRESSO_VERSION') || exit;

/**
 * This is used to generate an array that can be used to generate a schema for a given model.
 * The format for the generated array follows the structure given in the json-schema standard
 * @see http://json-schema.org
 *
 * @package    EventEspresso
 * @subpackage core\db_models\helpers
 * @author     Darren Ethier
 * @since      4.9.24.rc.018
 */
class ModelSchema
{
    /**
     * Return the schema for a given model from a given model.
     * @param \EEM_Base $model
     * @return array
     */
    public function getModelSchema(EEM_Base $model)
    {
        $fields_on_model = $model->field_settings();
        return $this->getModelSchemaForFields($fields_on_model, $model);
    }


    /**
     * Get the json schema for a given set of model fields.
     * @param \EE_Model_Field_Base[]     $model_fields
     * @param \EEM_Base $model
     * @return array
     */
    public function getModelSchemaForFields(array $model_fields, EEM_Base $model)
    {
        $schema = $this->getInitialSchemaStructure($model);
        foreach ($model_fields as $field => $model_field) {
            if (! $model_field instanceof EE_Model_Field_Base) {
                continue;
            }
            $schema['properties'][$field] = $model_field->get_schema();
        }
        return $schema;
    }


    /**
     * Outputs the schema header for a model.
     * @param \EEM_Base $model
     * @return array
     */
    protected function getInitialSchemaStructure(EEM_Base $model)
    {
        return array(
            '$schema' => 'http://json-schema.org/draft-04/schema#',
            'title' => $model->get_this_model_name(),
            'type' => 'object',
            'properties' => array()
        );
    }


    /**
     * Helper function that simply allows one to get whatever has been generated in an actual json object as opposed to
     * an array.
     * @param array $array_to_convert
     * @return bool|false|mixed|string
     */
    public function toJson($array_to_convert)
    {
        return wp_json_encode($array_to_convert);
    }
}
