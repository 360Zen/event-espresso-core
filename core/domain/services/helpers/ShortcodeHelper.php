<?php
namespace EventEspresso\core\domain\services\helpers;

use EE_Error;
use EE_Registry;
use EES_Shortcode;
use ReflectionClass;
use WP;

defined('EVENT_ESPRESSO_VERSION') || exit;



/**
 * Class Shortcodes
 * Description
 *
 * @package       Event Espresso
 * @author        Brent Christensen
 * @since         $VID:$
 */
class ShortcodeHelper
{

    /**
     * @type  EE_Registry $registry
     * @access    protected
     */
    protected $registry;



    /**
     * Shortcodes constructor.
     *
     * @param EE_Registry $registry
     */
    public function __construct(EE_Registry $registry)
    {
        $this->registry = $registry;
    }



    /**
     * get classname, remove EES_prefix, and convert to UPPERCASE
     *
     * @param string $class_name
     * @return string
     */
    public static function generateShortcodeTagFromClassName($class_name)
    {
        return strtoupper(str_replace('EES_', '', $class_name));
    }



    /**
     * add EES_prefix and Capitalize words
     *
     * @param string $tag
     * @return string
     */
    public static function generateShortcodeClassNameFromTag($tag)
    {
        // order of operation runs from inside to out
        // 5) maybe add prefix
        return ShortcodeHelper::addShortcodeClassPrefix(
            // 4) find spaces, replace with underscores
            str_replace(
                ' ',
                '_',
                // 3) capitalize first letter of each word
                ucwords(
                    // 2) also change to lowercase so ucwords() will work
                    strtolower(
                        // 1) find underscores, replace with spaces so ucwords() will work
                        str_replace(
                            '_',
                            ' ',
                            $tag
                        )
                    )
                )
            )
        );
    }



    /**
     * maybe add EES_prefix
     *
     * @param string $class_name
     * @return string
     */
    public static function addShortcodeClassPrefix($class_name)
    {
        return strpos($class_name, 'EES_') === 0 ? $class_name : 'EES_' . $class_name;
    }



    /**
     * @return array
     */
    public function getEspressoShortcodeTags()
    {
        static $shortcode_tags = array();
        if (empty($shortcode_tags)) {
            $shortcode_tags = array_keys((array)$this->registry->shortcodes);
        }
        return $shortcode_tags;
    }



    /**
     * @param string $content
     * @return string
     */
    public function doShortcode($content)
    {
        foreach ($this->getEspressoShortcodeTags() as $shortcode_tag) {
            if (strpos($content, $shortcode_tag) !== false) {
                $shortcode_class = ShortcodeHelper::generateShortcodeClassNameFromTag($shortcode_tag);
                $this->initializeShortcode($shortcode_class);
            }
        }
        return do_shortcode($content);
    }



    /**
     * given a shortcode name, will instantiate the shortcode and call it's run() method
     *
     * @param string $shortcode_class
     * @param WP    $wp
     */
    public function initializeShortcode($shortcode_class = '', WP $wp = null)
    {
        // don't do anything if shortcode is already initialized
        if ($this->registry->shortcodes->{$shortcode_class} instanceof EES_Shortcode){
            return;
        }
        // let's pause to reflect on this...
        $sc_reflector = new ReflectionClass(ShortcodeHelper::addShortcodeClassPrefix($shortcode_class));
        // ensure that class is actually a shortcode
        if (
            defined('WP_DEBUG')
            && WP_DEBUG === true
            && ! $sc_reflector->isSubclassOf('EES_Shortcode')
        ) {
            EE_Error::add_error(
                sprintf(
                    __(
                        'The requested %s shortcode is not of the class "EES_Shortcode". Please check your files.',
                        'event_espresso'
                    ),
                    $shortcode_class
                ),
                __FILE__,
                __FUNCTION__,
                __LINE__
            );
            add_filter('FHEE_run_EE_the_content', '__return_true');
            return;
        }
        global $wp;
        // and pass the request object to the run method
        $this->registry->shortcodes->{$shortcode_class} = $sc_reflector->newInstance();
        // fire the shortcode class's run method, so that it can activate resources
        $this->registry->shortcodes->{$shortcode_class}->run($wp);
    }

}
// End of file Shortcodes.php
// Location: EventEspresso\core\domain\services\helpers/Shortcodes.php