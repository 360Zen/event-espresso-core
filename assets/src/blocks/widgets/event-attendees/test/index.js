/**
 * External dependencies
 */
import { blockEditRender } from '../../../helpers';

/**
 * Internal dependencies
 */
import { name, settings } from '../';

describe ( name, () => {
	test ( 'block edit matches snapshot', () => {
		const wrapper = blockEditRender ( name, settings );
		expect ( wrapper ).toMatchSnapshot ();
	} );
} );

// location: /assets/src/blocks/widgets/event-attendees/test/index.js