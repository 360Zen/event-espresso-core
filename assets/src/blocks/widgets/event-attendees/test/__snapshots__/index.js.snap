// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`eventespresso/widgets-event-attendees block edit matches snapshot 1`] = `
<[object Object]
  value={
    Object {
      "clientId": undefined,
      "focusedElement": null,
      "isSelected": false,
      "name": "eventespresso/widgets-event-attendees",
      "setFocusedElement": [Function],
    }
  }
>
  <WithFilters(Edit)
    attributes={
      Object {
        "datetimeId": 0,
        "displayOnArchives": false,
        "eventId": 0,
        "limit": 10,
        "showGravatar": false,
        "status": "RAP",
        "ticketId": 0,
      }
    }
    isSelected={false}
    name="eventespresso/widgets-event-attendees"
    setAttributes={[Function]}
    user={Object {}}
  />
</[object Object]>
`;
